metaflow/plugins/metadata/service.py

```diff
@@ -73,7 +94,55 @@ def register_task_id(self,
                            sys_tags=sys_tags)
         finally:
             self._register_code_package_metadata(run_id, step_name, task_id)
-        
+
+    def _start_heartbeat(self, heartbeat_type, flow_id, run_id, step_name=None, task_id=None):
+        if self._already_started():
+            raise Exception("heartbeat already started")
```

Comment:
With the way metaflow is designed it's not really possible for the same ServiceMetadataProvider object to be shared by both a run and a task since they already execute in different sub-processes. We can avoid raising an exception but that would be a departure from the internal implementation

Reply:
Yes, I agree that it is is not possible (and shouldn't be). The point I was making is that this is not really documented anywhere and if you just look at the code as a casual observer, you would be wondering why there are two places to call start and an exception if you call it twice :). For the exception part, my point was that this is an optional feature (sidecar) and can kill your main process for something that you really can't do anything about.

Reply:
I added a few comments to hopefully clarify things. We don't have any developer documentation as far as I know in OSS so in code comments will have to do.

Re: the exception: I totally get your point and we went back and forth on this in the original implementation a couple years ago. The reasoning behind throwing a hard error here is because logically in the code this situation should never happen and if it did it means data in the metadata service could become corrupted and we want to terminate the run and fail loudly. If the heartbeat fails to start or dies part way through the execution of the main process it's considered a soft failure and we opt to fail silently.

There are arguments to be made in either direction but that is the call we took back in 2019. We can revisit that now but we probably need to think through the failure scenarios again.